lib/cchain/provider/abci.go

```diff
@@ -280,3 +315,194 @@ func (a rpcAdaptor) Invoke(ctx context.Context, method string, req, resp any, _
 func spanName(endpoint string) string {
 	return "cprovider/" + endpoint
 }
+
+// Finds the height at which the requested offset can be found in state.
+func searchOffsetInHistory(ctx context.Context, client ABCIClient, cl atypes.QueryClient, chainVer xchain.ChainVersion, fromOffset uint64) (int, error) {
+	// TODO: Aayush: Is it fine to be capturing a second thing here?
+	const endpoint = "search_offset"
+	defer latency(endpoint)
+
+	// Exponentially backoff to find a good start point for binary search, this prefers more recent queries
+	info, err := client.ABCIInfo(ctx)
+	if err != nil {
+		return -1, errors.Wrap(err, "abci query info")
+	}
+	startHeightIndex := 0
+	endHeightIndex := int(info.Response.LastBlockHeight)
+	lookback := 1
+	lookbackStepsCounter := 0 // For metrics only
+	queryHeight := endHeightIndex - lookback
+	for queryHeight > 0 {
+		lookbackStepsCounter++
+		earliestAtt, ok, err := queryEarliestAttestation(ctx, cl, chainVer, queryHeight)
+		if IsErrHistoryPruned(err) {
+			// We've jumped to before the prune height, but _might_ still have the requested offset
+			earliestStoreHeight, err := getEarliestStoreHeight(ctx, cl, chainVer, queryHeight+1)
+			if err != nil {
+				return -1, errors.Wrap(err, "failed to get earliest store height")
+			}
+			earliestAtt, ok, err = queryEarliestAttestation(ctx, cl, chainVer, earliestStoreHeight)
+			if err != nil {
+				incQueryErr(endpoint)
+				return -1, errors.Wrap(err, "abci query earliest-attestation-in-state")
+			}
+
+			// If we're so far back that no attestation is found, or that we're before fromOffset,
+			// that's a good breaking point for binary search
+			if !ok || earliestAtt.BlockHeader.BlockOffset <= fromOffset {
+				startHeightIndex = earliestStoreHeight
+				break
+			}
+
+			// Otherwise, we just don't have the needed state, fail
+			return -1, errors.New("missing state for requested offset")
+		}
+		if err != nil {
+			incQueryErr(endpoint)
+			return -1, errors.Wrap(err, "abci query earliest-attestation-in-state")
+		}
+
+		// If we're before the first attestation, or found an earlier attestation, it's a good start height
+		if !ok || earliestAtt.BlockHeader.BlockOffset <= fromOffset {
+			startHeightIndex = queryHeight
+			break
+		}
+
+		// Otherwise, keep moving back
+		endHeightIndex = queryHeight
+		lookback *= 2
+		if queryHeight > lookback {
+			queryHeight -= lookback
+		} else {
+			queryHeight = 1
+		}
+	}
+
+	// We now have reasonable start and end indices for binary search
+	binarySearchStepsCounter := 0 // For metrics only
```

Comment:
we prefer to initialise zero variables using the `var foo int` pattern. As per [Practical Go](https://dave.cheney.net/practical-go/presentations/gophercon-singapore-2019.html#_use_a_consistent_declaration_style)